### PR TITLE
Miken/makepeoplelikemeforwritingtests

### PR DESCRIPTION
**Pull request by digitalbuddha**
ScopesConstructionTests.kt  is the bulk of it, notice how easy it is to swap dependencies within your tests.

### CODE REVIEW

File: integration-tests/src/androidTest/java/com/dropbox/kaiken/skeleton/scoping/ScopesConstructionTests.kt [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    @Component.Factory
+    interface Factory {
+        fun create(
+//            @BindsInstance app: SkeletonOwner,
+        ): SkeletonTestComponent
+    }
+}
+
+@ContributesBinding(SkeletonScope::class, replaces = [KaikenConfig::class])
+class FakeSkeletonConfig @Inject constructor() : SkeletonConfig {
```
Comment (digitalbuddha) [inline integration-tests/src/androidTest/java/com/dropbox/kaiken/skeleton/scoping/ScopesConstructionTests.kt:57]: I looooove anvil. We have a binding that is contributed somewhere within skeleton main source. In tests, I can change this binding. One of the ways I can change it is by making a fake SkeletonConfig which no longer depends on an application class. Next, I can make the `SkeletonTestComponent` which never binds the application instance in. This way you can have variants of the skeleton graph which are decoupled from android. 

Similar check out the `FakeActiveUserManager` below which drops the real dependency on coroutines for this noop version

File: integration-tests/build.gradle [hunk 2]
```diff
@@ -26,6 +28,9 @@ dependencies {
     androidTestImplementation deps.truth
 
     androidTestImplementation project(':testing')
+    kaptAndroidTest deps.dagger2_compiler
+
```
Comment (chris-mitchell) [inline integration-tests/build.gradle:32]: nit: whitespace

File: integration-tests/src/androidTest/java/com/dropbox/kaiken/skeleton/scoping/ScopesConstructionTests.kt [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    }
+}
+
+@MergeComponent(SkeletonScope::class)
+@SingleIn(SkeletonScope::class)
+interface SkeletonTestComponent : SdkSpec {
+    @Component.Factory
+    interface Factory {
+        fun create(
+//            @BindsInstance app: SkeletonOwner,
```
Comment (chris-mitchell) [inline integration-tests/src/androidTest/java/com/dropbox/kaiken/skeleton/scoping/ScopesConstructionTests.kt:53]: unused?

File: integration-tests/src/androidTest/java/com/dropbox/kaiken/skeleton/scoping/ScopesConstructionTests.kt [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                    .appComponent()
+
+            override fun createUserServices(
+                appServices: AppServices,
+                user: SkeletonUser
+            ): UserServices =
+                appServices
+                    .cast<UserParentComponent>()
+                    .createUserComponent()
+                    .userComponent(user.userId.toInt())
```
[the remark this answers is not in the document]
Comment (digitalbuddha) [reply integration-tests/src/androidTest/java/com/dropbox/kaiken/skeleton/scoping/ScopesConstructionTests.kt:74]: yup will follow up, this is just to get structure right for components

File: integration-tests/src/androidTest/java/com/dropbox/kaiken/skeleton/scoping/ScopesConstructionTests.kt [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+class NoOpUserTeardownHelper @Inject constructor() : UserTeardownHelper {
+    override fun teardown() {
+        // No op
+    }
+}
+
+@ContributesTo(AuthRequiredActivityScope::class)
+interface UserTeardownHelperProvider {
+    fun userTeardownHelper(): UserTeardownHelper
+//    fun appTeardownHelper():AppTeardownHelper
```
[the remark this answers is not in the document]
Comment (digitalbuddha) [reply integration-tests/src/androidTest/java/com/dropbox/kaiken/skeleton/scoping/ScopesConstructionTests.kt:114]: intentional, discussed offline

**Review by chris-mitchell — APPROVED** [review]

**Comment by digitalbuddha** [comment]
bah lint, fixing main now

**Pull request merged**